src/test/fuzz/banman.cpp

```diff
@@ -68,12 +68,17 @@ FUZZ_TARGET(banman, .init = initialize_banman)
             CallOneOf(
                 fuzzed_data_provider,
                 [&] {
-                    ban_man.Ban(ConsumeNetAddr(fuzzed_data_provider),
-                                ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                    const std::string name{fuzzed_data_provider.ConsumeRandomLengthString(512)};
+                    if (const std::optional<CNetAddr>& addr{LookupHost(name, /*fAllowLookup=*/false)}) {
+                        ban_man.Ban(addr.value(), ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                    }
                 },
                 [&] {
-                    ban_man.Ban(ConsumeSubNet(fuzzed_data_provider),
-                                ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                    const std::string name{fuzzed_data_provider.ConsumeRandomLengthString(512)};
```

Comment:
> Does the string representation make it harder for the fuzz engine to reach coverage?

Hm, yes, perhaps an alternative to facilitate it (didnt test yet, just wondering):
```diff
+        CSubNet subnet;
+        CNetAddr net_addr;
         LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 300)
         {
             CallOneOf(
                 fuzzed_data_provider,
                 [&] {
-                    const std::string name{fuzzed_data_provider.ConsumeRandomLengthString(512)};
-                    if (const std::optional<CNetAddr>& addr{LookupHost(name, /*fAllowLookup=*/false)}) {
-                        ban_man.Ban(addr.value(), ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                    CSubNet new_subnet{ConsumeSubNet(fuzzed_data_provider)};
+                    if (LookupSubNet(new_subnet.ToString(), new_subnet)) {
+                        subnet = new_subnet;
                     }
                 },
                 [&] {
-                    const std::string name{fuzzed_data_provider.ConsumeRandomLengthString(512)};
-                    CSubNet subnet;
-                    if (LookupSubNet(name, subnet)) {
-                        ban_man.Ban(subnet, ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                    CNetAddr new_netaddr{ConsumeNetAddr(fuzzed_data_provider)};
+                    if (const std::optional<CNetAddr>& addr{LookupHost(new_netaddr.ToStringAddr(), /*fAllowLookup=*/false)}) {
+                        net_addr = addr.value();
                     }
                 },
+                [&] {
+                    ban_man.Ban(net_addr, ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                },
+                [&] {
+                    ban_man.Ban(subnet, ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
+                },
                 [&] {
                     ban_man.ClearBanned();
                 },
                 [&] {
-                    ban_man.IsBanned(ConsumeNetAddr(fuzzed_data_provider));
+                    ban_man.IsBanned(net_addr);
                 },
                 [&] {
-                    ban_man.IsBanned(ConsumeSubNet(fuzzed_data_provider));
+                    ban_man.IsBanned(subnet);
                 },
                 [&] {
-                    ban_man.Unban(ConsumeNetAddr(fuzzed_data_provider));
+                    ban_man.Unban(net_addr);
                 },
                 [&] {
-                    ban_man.Unban(ConsumeSubNet(fuzzed_data_provider));
+                    ban_man.Unban(subnet);
                 },
                 [&] {
                     banmap_t banmap;
@@ -103,7 +110,7 @@ FUZZ_TARGET(banman, .init = initialize_banman)
                     ban_man.DumpBanlist();
                 },
                 [&] {
-                    ban_man.Discourage(ConsumeNetAddr(fuzzed_data_provider));
+                    ban_man.Discourage(net_addr);
                 });
         }
```

Reply:
Would it be possible to still call `ConsumeSubNet()` but check if the returned `CSubNet` object `IsValid()` before feeding it to `Ban()`?

`ConsumeSubNet()` could generate nonsensical stuff like `2bqghnldu6mcug4pikzprwhtjjnsyederctvci6klcwzepnjd46ikjyd.onion/16` which is later an invalid `CSubNet` whose constructor only assumes IPv4 or IPv6 subnets valid. This is the reason also for https://github.com/bitcoin/bitcoin/pull/27071#issuecomment-1774139873 (it generates `fc00::%538976288/6` IPv6 which later gets converted to CJDNS and rightfully turns out to be an invalid subnet because it is not IPv4 or IPv6).

Or change ``ConsumeSubNet()` to only produce valid stuff?

Reply:
https://github.com/bitcoin/bitcoin/pull/27935#discussion_r1368594365 still could call `Ban()` with an invalid address or subnet (default constructed is `!IsValid()`). I am not sure if we should be doing that from the tests:
* No code should be doing this, but still some buggy code may do
* It is good to exercise that and see that `BanMan` does not crash
* `BanMan` removes invalid entries when re-reading the bans from the json file, so it is unrealistic to expect that the ban entries will be the same before and after read from the json file
* Further, this looks like an inconsistency in `BanMan` that it accepts and invalid entry into its map via `Ban()`, but rejects invalid entries from the json file. Maybe `BanMan` should reject invalid entries also from the `Ban()` method.

Reply:
I think the problem is the logic is all in the RPC (lookup calls and the vailidity check). We could move it to a function.

Reply:
This is still an issue in the latest version (8adbd44c9b). It seems unlikely that the fuzzer will generate a string that makes `LookupHost()` happy. I think it is best if this test keeps using `ConsumeNetAddr()` and `ConsumeSubNet()` but those functions should return valid/sensible values. Something like:

```cpp
inline CSubNet ConsumeSubNet(FuzzedDataProvider& fuzzed_data_provider) noexcept
{
    const CNetAddr net_base = ConsumeNetAddr(fuzzed_data_provider, {NET_IPV4, NET_IPV6}, /*always_valid=*/true);
    const uint8_t cidr_max = net_base.IsIPv4() ? ADDR_IPV4_SIZE : ADDR_IPV6_SIZE;
    return {net_base, fuzzed_data_provider.ConsumeIntegralInRange<uint8_t>(0, cidr_max * 8)};
}
```

Reply:
Yes, this was my doubt too. But then it is unrealistic to expect that banman ser/deser will produce identical result as the input. Another try - what about feeding it anything (including invalid stuff), but keeping track if we ever fed it an invalid address or subnet and if yes, then don't do the assert at the end `assert(banmap == banmap_read);`?

Reply:
> Yes, this was my doubt too. But then it is unrealistic to expect that banman ser/deser will produce identical result as the input. Another try - what about feeding it anything (including invalid stuff), but keeping track if we ever fed it an invalid address or subnet and if yes, then don't do the assert at the end assert(banmap == banmap_read);?

sgtm as well. Can I change the approach so?